src/main/java/org/alliancegenome/curation_api/services/GeneExpressionAnnotationService.java

```diff
@@ -0,0 +1,135 @@
+
+	@Inject	ExpressionAnnotationDAO expressionAnnotationDAO;
+
+	@Inject	GeneExpressionAnnotationDAO geneExpressionAnnotationDAO;
+
+	@Inject GeneService geneService;
+
+	@Inject MmoTermService mmoTermService;
+
+	@Inject	ReferenceService referenceService;
+
+	@Inject GeneExpressionAnnotationUniqueIdHelper geneExpressionAnnotationUniqueIdHelper;
+
+	@Inject	GeneExpressionAnnotationFmsDTOValidator geneExpressionAnnotationFmsDTOValidator;
+
+	@Inject DataProviderService dataProviderService;
+
+	@Inject VocabularyTermService vocabularyTermService;
+
+	@Inject	PersonDAO personDAO;
+
+	@Inject PersonService personService;
+
+	@Override
+	protected void init() {
+		setSQLDao(geneExpressionAnnotationDAO);
+	}
+
+	@Override
+	public GeneExpressionAnnotation deprecateOrDeleteAnnotationAndNotes(Long id, Boolean throwApiError, String loadDescription, Boolean deprecate) {
+		GeneExpressionAnnotation geneExpressionAnnotation = geneExpressionAnnotationDAO.find(id);
+
+		if (geneExpressionAnnotation == null) {
+			String errorMessage = "Could not find Gene Expression Annotation with id: " + id;
+			if (throwApiError) {
+				ObjectResponse<GeneExpressionAnnotation> response = new ObjectResponse<>();
+				response.addErrorMessage("id", errorMessage);
+				throw new ApiErrorException(response);
+			}
+			log.error(errorMessage);
+			return null;
+		}
+		if (deprecate) {
+			if (!geneExpressionAnnotation.getObsolete()) {
+				geneExpressionAnnotation.setObsolete(true);
+				if (authenticatedPerson.getId() != null) {
+					geneExpressionAnnotation.setUpdatedBy(personDAO.find(authenticatedPerson.getId()));
+				} else {
+					geneExpressionAnnotation.setUpdatedBy(personService.fetchByUniqueIdOrCreate(loadDescription));
+				}
+				geneExpressionAnnotation.setDateUpdated(OffsetDateTime.now());
+				return geneExpressionAnnotationDAO.persist(geneExpressionAnnotation);
+			} else {
+				return geneExpressionAnnotation;
+			}
+		} else {
+			geneExpressionAnnotationDAO.remove(id);
+		}
+		return null;
+	}
+
+	public List<Long> getAnnotationIdsByDataProvider(BackendBulkDataProvider dataProvider) {
+		Map<String, Object> params = new HashMap<>();
+		params.put(EntityFieldConstants.DATA_PROVIDER, dataProvider.sourceOrganization);
+		if (StringUtils.equals(dataProvider.sourceOrganization, "RGD") || StringUtils.equals(dataProvider.sourceOrganization, "XB")) {
+			params.put(EntityFieldConstants.PA_SUBJECT_TAXON, dataProvider.canonicalTaxonCurie);
+		}
+		List<Long> annotationIds = geneExpressionAnnotationDAO.findIdsByParams(params);
+		return annotationIds;
+	}
+
+	@Transactional
+	@Override
+	public GeneExpressionAnnotation upsert(GeneExpressionFmsDTO geneExpressionFmsDTO, BackendBulkDataProvider dataProvider) throws ObjectUpdateException {
+		if (geneExpressionAnnotationFmsDTOValidator.isExpressionAnnotationValid(geneExpressionFmsDTO)) {
```

Comment:
Umm... this looks like to me if the annotation is valid we are going to go create a new one anyway. I think the logic should be that you generate the unique ID and see if that is in the database first before creating a new one. @markquintontulloch maybe I am wrong here?

Reply:
@oblodgett - I think that's what is going on here.  A new annotation is created, but that is switched for the annotation from the DB if one is found based on the uniqueId generated from the DTO.  However, I think it would be cleaner to not create (and discard) a new annotation unless it is not found in the DB.

Reply:
But the logic here is that when `isExpressionAnnotationValid` returns true then the annotation is created. I would at least except a `!isExpressionAnnotationValid` then create the new one.

Reply:
Changed and moved to validation